CodeEdit/Features/Tasks/CEActiveTask.swift

```diff
@@ -0,0 +1,229 @@
+//
+//  CEActiveTask.swift
+//  CodeEditTests
+//
+//  Created by Tommy Ludwig on 24.06.24.
+//
+
+import SwiftUI
+import Combine
+
+/// Stores the state of a task once it's executed
+class CEActiveTask: ObservableObject, Identifiable, Hashable {
+    /// The current progress of the task.
+    @Published private(set) var output: String  = ""
+
+    /// The status of the task.
+    @Published private(set) var status: CETaskStatus = .notRunning
+
+    /// The name of the associated task.
+    @ObservedObject var task: CETask
+
+    var process: Process?
+    var outputPipe: Pipe?
+
+    private var cancellables = Set<AnyCancellable>()
+
+    init(task: CETask) {
+        self.task = task
+        self.process = Process()
+        self.outputPipe = Pipe()
+
+        self.task.objectWillChange.sink { _ in
+            self.objectWillChange.send()
+        }.store(in: &cancellables)
+    }
+
+    func run() {
+        guard let process, let outputPipe else { return }
+
+        Task { await updateTaskStatus(to: .running) }
+        createStatusTaskNotification()
+
+        process.terminationHandler = { [weak self] _ in
+            self?.handleProcessFinished()
+        }
+
+        Task.detached {
+            outputPipe.fileHandleForReading.readabilityHandler = { fileHandle in
+                let data = fileHandle.availableData
+                Task {
+                    await self.updateOutput(String(decoding: data, as: UTF8.self))
+                }
+            }
+
+            do {
+                try await TaskShell.executeCommandWithShell(
+                    process: process,
+                    command: self.task.fullCommand,
+                    environmentVariables: self.task.environmentVariablesDictionary,
+                    shell: TaskShell.zsh, // TODO: Let user decide which shell he uses
+                    outputPipe: outputPipe
+                )
+            } catch { print(error) }
+        }
+    }
+
+    func handleProcessFinished() {
```

Comment:
This could be made into an async method, it's only called once in `run` and could be added to a task in the termination handler. It would simplify this function a little as you wouldn't deal with as many optionals or dispatched tasks.

Reply:
That makes the function way easier to read—thanks!